include/rotatingMHD/solver_class.h

```diff
@@ -0,0 +1,408 @@
+ *    \quad \mathcal{L}(\bs{a}_\textrm{s}) = \bs{f}, \qquad
+ *    \forall (\bs{x}, t) \in \Omega \times \mathbb{R}_0
+ * \f]
+ * where \f$ \bs{a}_\textrm{s} \f$ is a vector, on which a divergence-free
+ * constraint is enforced, i.e. a solenoidal vector;
+ * \f$ \mathcal{L} \f$ is a differential operator;
+ * \f$ \bs{x} \f$ is the position vector;
+ * \f$ t \f$ is the time;
+ * \f$ \Omega \f$ is the spatial domain;
+ * \f$ \mathbb{R}_0 \f$ is the temporal domain, all real numbers
+ * including zero. Based on a temporal decoupling of variables and
+ * the \textsc{Helmholtz} decomposition of an arbitrary vector into
+ * the sum of a solenoidal and a irrotational vector
+ * \f[
+ *    \bs{a} = \bs{a}_\textrm{s} + \nabla \phi,
+ * \f]
+ * the projection scheme decouples \f$ \mathcal{L}(\bs{a}_\textrm{s})\f$
+ * from the divergence-free condition by solving a diffusion step
+ * \f[
+ *    \mathcal{L}(\bs{a}) = \bs{f};
+ * \f]
+ * a projection step, in which \f$ \bs{a} \f$ is projected
+ * into the space of solenoidal vectors, i.e., \f$\bs{a}_\textrm{s}\f$
+ * is computed; and a correction step, which updates \f$ \phi \f$ to
+ * the current time step. A more detailed explanation, can be found in
+ * the child classes of @ref ProjectionSolverBase.
+ * @tparam dim An integer indicating the spatial dimension of the
+ * problem.
+ * @attention The solenoidal vector field is eliminated in the internal
+ * algorithm. Therefore, the computed vector field is not solenoidal but
+ * it fulfills the boundary conditions.
+ */
+template<int dim>
+class ProjectionSolverBase : SolverBase<dim>
+{
+public:
+
+/**
+ * @brief Construct a new ProjectionSolverBase instance
+ *
+ * @param time_stepping See the documentation for @ref SolverBase
+ * @param vector The vector field on which the divergence-free condition
+ * is imposed
+ * @param lagrange_multiplier The lagrange multiplier imposing the
+ * divergence-free condition
+ * @param external_mapping See the documentation for @ref SolverBase
+ * @param external_pcout See the documentation for @ref SolverBase
+ * @param external_timer See the documentation for @ref SolverBase
+ */
+ProjectionSolverBase(
+  TimeDiscretization::VSIMEXMethod                &time_stepping,
+  const std::shared_ptr<Mapping<dim>>             external_mapping =
+    std::shared_ptr<Mapping<dim>>(),
+  const std::shared_ptr<ConditionalOStream>       external_pcout =
+    std::shared_ptr<ConditionalOStream>(),
+  const std::shared_ptr<TimerOutput>              external_timer =
+    std::shared_ptr<TimerOutput>());
+
+
+/**
+ * @brief The auxiliary scalar field introduced to simplify the
+ * diffusion and projection step.
+ *
+ */
+std::shared_ptr<Entities::FE_ScalarField<dim>>  auxiliary_scalar;
+
+/**
+ * @brief Sets the internal supply term shared pointer equal to
+ * @ref supply_term.
+ *
+ * @param supply_term The external supply term pointer.
+ */
+void set_supply_term(
+  std::shared_ptr<TensorFunction<1, dim>> supply_term);
```

Comment:
Couldn't we use reference to a `TensorFunction<1,dim>` as an input argument instead? This would simplify application because we don't have to mess around the `shared_ptr` objects. See https://godbolt.org/z/vb7vYd3s9 for an example.

Reply:
Not sure what is to be demonstrated by the example. The internal member related to the supply term has to be a pointer. I have to admit that I do not have any arguments for the use of shared pointer instead of raw pointers. But since we have been using shared pointers throughout the library so far, I also chose a shared pointer here for consistency. Nonetheless, I changed it to mirror the internal treatment of the supply terms that we have been using up until now (See 085049b)

Reply:
> Not sure what is to be demonstrated by the example. The internal member related to the supply term has to be a pointer. I have to admit that I do not have any arguments for the use of shared pointer instead of raw pointers. But since we have been using shared pointers throughout the library so far, I also chose a shared pointer here for consistency. Nonetheless, I changed it to mirror the internal treatment of the supply terms that we have been using up until now (See [085049b](https://github.com/j507/RotatingMHD/commit/085049b260d5c597d34d02062ac71b421ae7134b))

The only point of using a reference is that you can write more readable application code. Instead of
```c++
{
  std::shared_ptr<TensorFunction<1, dim>> function_ptr = std::make_shared<MySupplyTerm<dim>>(/* ... */);
  navier_stokes_solver.set_supply_term(function_ptr);
}
```
we could simply write
```c++
{
  MySupplyTerm<dim> supply_term(/* ... */);
  navier_stokes_solver.set_supply_term(supply_term);
}
```